src/github.com/matrix-org/dendrite/roomserver/query/query.go

```diff
@@ -418,6 +424,43 @@ func (r *RoomserverQueryAPI) QueryServerAllowedToSeeEvent(
 	return nil
 }
 
+// QueryReserveRoomID implements api.RoomserverQueryAPI
+func (r *RoomserverQueryAPI) QueryReserveRoomID(
+	ctx context.Context,
+	request *api.QueryReserveRoomIDRequest,
+	response *api.QueryReserveRoomIDResponse,
+) error {
+	nid, err := r.DB.RoomNID(ctx, request.RoomID)
+	if err != nil && err != sql.ErrNoRows {
+		return err
+	}
+
+	if nid != 0 {
+		response.Success = false
+		return nil
+	}
+
+	// Check if Room ID has already been reserved.
+	reservedSince, err := r.DB.RoomIDReserved(ctx, request.RoomID)
+	if reservedSince != (time.Time{}) {
+		response.Success = false
+		return nil
+	}
+
+	if err != nil && err != sql.ErrNoRows {
+		return err
+	}
+
+	err = r.DB.ReserveRoomID(ctx, request.RoomID)
+	if err != nil {
+		return err
+	}
+
+	response.Success = true
+
+	return nil
```

Comment:
All of this needs to happen in a single database transaction to avoid races where two reserve room requests come in at the same time. In fact, I would probably have a single table which was just a list of reserved/used roomIDs and do:

`INSERT INTO roomserver_reserved_rooms (room_id) VALUES ($1) ON CONFLICT (room_id) DO NOTHING`

and then check if that returns any rows updated or not. If no rows were updated then we know that the roomID has already been reserved.